File: forms/organ-donation-assessment/back-end-with-loco/src/engine/donation_grader.rs
//! Donation grader module.

use super::donation_rules::all_rules;
use super::flagged_issues::detect_additional_flags;
use super::types::{AssessmentData, Eligibility, FiredRule, GradingResult, RiskLevel};

/// Pure function: grades an Organ Donation Assessment.
///
/// Returns a [`GradingResult`] containing:
/// - `eligibility` — assessor decision (if recorded) or engine-suggested decision
/// - `risk_level` — engine-derived risk classification
/// - `suggested_eligibility` — engine-derived suitability
/// - fired rules (one per matched donor rule)
/// - additional flags (real-time safety alerts)
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = evaluate_rules(data);
    let additional_flags = detect_additional_flags(data);

    let (eligibility, risk_level, suggested_eligibility) =
        classify_eligibility(&fired_rules, data);

    GradingResult {
        eligibility,
        risk_level,
        suggested_eligibility,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

/// Evaluate every donor rule against the data, returning the fired rules.
pub fn evaluate_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut fired = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                grade: rule.grade,
            });
        }
    }
    fired
}

/// Classify eligibility + risk from fired rules and the recorded decision.
///
/// Any single Grade-4 finding (other than the "ideal donor" Grade-1 markers
/// such as SU-001) → unsuitable / critical.
/// Any Grade-3 finding → conditionally-suitable / high.
/// Any Grade-2 finding → conditionally-suitable / moderate.
/// Otherwise → suitable / low.
///
/// When the assessor explicitly recorded an eligibility decision in step 10
/// it wins for the displayed eligibility, but the engine-derived risk level
/// is preserved.
pub fn classify_eligibility(
    fired_rules: &[FiredRule],
    data: &AssessmentData,
) -> (Eligibility, RiskLevel, Eligibility) {
    // Ignore the explicit "ideal donor" Grade-1 markers when counting penalties.
    let positive_ideal_ids = ["SU-001"];

    let flagged: Vec<&FiredRule> = fired_rules
        .iter()
        .filter(|r| !positive_ideal_ids.contains(&r.id.as_str()))
        .collect();

    let grade4 = flagged.iter().filter(|r| r.grade == 4).count();
    let grade3 = flagged.iter().filter(|r| r.grade == 3).count();
    let grade2 = flagged.iter().filter(|r| r.grade == 2).count();

    let (suggested_eligibility, risk_level): (&str, &str) = if grade4 > 0 {
        ("unsuitable", "critical")
    } else if grade3 > 0 {
        ("conditionally-suitable", "high")
    } else if grade2 > 0 {
        ("conditionally-suitable", "moderate")
    } else {
        ("suitable", "low")
    };

    let assessor_decision = &data.eligibility_allocation.eligibility_decision;
    let eligibility = if assessor_decision.is_empty() {
        suggested_eligibility.to_string()
    } else {
        assessor_decision.clone()
    };

    (
        eligibility,
        risk_level.to_string(),
        suggested_eligibility.to_string(),
    )
}

/// Backwards-compat helper for tests.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    evaluate_rules(data)
}

File: forms/organ-donation-assessment/front-end-with-svelte/src/lib/engine/donation-grader.ts
import type { AssessmentData, Eligibility, FiredRule, GradingResult, RiskLevel } from './types';
import { donationRules } from './donation-rules';
import { detectAdditionalFlags } from './flagged-issues';

/**
 * Evaluate every declarative donor rule against the given assessment data.
 * Returns the list of fired rules.
 */
function evaluateRules(data: AssessmentData): FiredRule[] {
	const fired: FiredRule[] = [];
	for (const rule of donationRules) {
		try {
			if (rule.evaluate(data)) {
				fired.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					grade: rule.grade
				});
			}
		} catch (e) {
			// Rule evaluation failed — log for debugging but continue grading.
			console.warn(`Donor rule ${rule.id} evaluation failed:`, e);
		}
	}
	return fired;
}

/**
 * Compute final eligibility + risk level from the fired rules and (where
 * recorded) the assessor's eligibility decision in step 10. When the assessor
 * explicitly recorded a decision, that decision wins for the displayed
 * eligibility, but the engine-derived risk level is preserved.
 *
 * The explicit "ideal donor" Grade-1 markers (e.g. SU-001 ASA-I) are excluded
 * from penalty counting, similar to the bone-marrow grader pattern.
 */
function classifyEligibility(
	firedRules: FiredRule[],
	data: AssessmentData
): { eligibility: Eligibility; suggestedEligibility: Eligibility; riskLevel: RiskLevel } {
	const positiveIdealIds = new Set(['SU-001']);

	const flagged = firedRules.filter((r) => !positiveIdealIds.has(r.id));
	const grade4 = flagged.filter((r) => r.grade === 4).length;
	const grade3 = flagged.filter((r) => r.grade === 3).length;
	const grade2 = flagged.filter((r) => r.grade === 2).length;

	let suggestedEligibility: Eligibility = 'suitable';
	let riskLevel: RiskLevel = 'low';

	if (grade4 > 0) {
		suggestedEligibility = 'unsuitable';
		riskLevel = 'critical';
	} else if (grade3 > 0) {
		suggestedEligibility = 'conditionally-suitable';
		riskLevel = 'high';
	} else if (grade2 > 0) {
		suggestedEligibility = 'conditionally-suitable';
		riskLevel = 'moderate';
	}

	const assessorDecision = data.eligibilityAllocation.eligibilityDecision;
	const eligibility: Eligibility = assessorDecision || suggestedEligibility;

	return { eligibility, suggestedEligibility, riskLevel };
}

/**
 * Pure function: evaluates all donor rules against assessment data and returns
 * the eligibility classification, risk level, fired rules, and flagged issues.
 */
export function gradeDonor(data: AssessmentData): GradingResult {
	const firedRules = evaluateRules(data);
	const { eligibility, suggestedEligibility, riskLevel } = classifyEligibility(firedRules, data);
	const additionalFlags = detectAdditionalFlags(data);

	return {
		eligibility,
		suggestedEligibility,
		riskLevel,
		firedRules,
		additionalFlags,
		timestamp: new Date().toISOString()
	};
}
